### PR TITLE
Add IMap.deleteAsync

### PR DESCRIPTION
**Pull request by vladykin**
Included basic tests, similar to those for removeAsync.

With new method added, MapProxySupport grew to 1502 lines, so had to raise the file size limit in checkstyle.xml.

Fixes https://github.com/hazelcast/hazelcast/issues/10702

Based on the abandoned https://github.com/hazelcast/hazelcast/pull/17739

Checklist:
- [ ] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [ ] Label `Add to Release Notes` or `Not Release Notes content` set
- [ ] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases
- [ ] New public APIs have `@Nonnull/@Nullable` annotations
- [ ] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

**Comment by mdumandag** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java:754]: The remove variant uses `invokeOnPartitionAsync`, can you also do the same here?

File: hazelcast/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTest.java
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTest.java:311]: Can you also please add tests for the server-side proxy? You can take a look at the tests that use removeAsync to see where to add tests to

Comment (vladykin) [reply hazelcast/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTest.java:311]: I looked but haven't found. There is MapProxyImplTest without any test cases for removeAsync. And there seems to be no test class for NearCachedMapProxyImpl.

Comment (vladykin) [reply hazelcast/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTest.java:311]: Could you please give me a more specific pointer, where to look?

Comment (mdumandag) [reply hazelcast/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTest.java:311]: When I looked at the usages of IMap#removeAsync, I see the tests. (I am using Intellij Idea). 

Here is the screenshot of the usage, you can check these tests and write equivalent tests if needed for the deleteAsync
image

Comment (vladykin) [reply hazelcast/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTest.java:311]: OK, I will work on adding more tests next week

Comment (vladykin) [reply hazelcast/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTest.java:311]: com.hazelcast.client.map
* ClientMapBasicTest - added testDeleteAsync, testDeleteAsync_whenKeyNotPresent
* ClientMapTest - added testAsyncDelete
* MapPreconditionsTest - added testDeleteAsync


com.hazelcast.client.map.impl.nearcache
* ClientMapNearCacheTest - added testAfterDeleteAsyncNearCacheIsInvalidated

com.hazelcast.core.standalone
* AllTest - added deleteAsync operation to loadMapOperations()

com.hazelcast.map
* AsyncTest - added testDeleteAsync, testDeleteAsyncWithImmediateTimeout, testDeleteAsyncWithNonExistentKey
* BasicMapTest - added testGetPutDeleteAsync, extended testNullChecks
* LocalMapStatsTest - added testDeleteAsync
* MapPreconditionsTest - added testDeleteAsync

com.hazelcast.map.impl.mapstore
* EntryLoaderSimpleTest - added testDeleteAsync_returnsNull
* EntryStoreSimpleTest - added testDeleteAsync

com.hazelcast.map.impl.nearcache
* AbstractMapNearCacheLocalInvalidationTest - added testDeleteAsync

com.hazelcast.splitbrainprotection.map
* MapSplitBrainProtectionWriteTest - added deleteAsync_successful_whenSplitBrainProtectionSize_met, deleteAsync_failing_whenSplitBrainProtectionSize_met

Comment (vladykin) [reply hazelcast/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTest.java:311]: Didn't add anything to DistributedDatastructuresMetricsTest and EventQueuePluginTest in com.hazelcast.internal.diagnostics. removeAsync is used in a long test scenario checking other subsystems. Not sure if it's appropriate to duplicate the whole test scenario to use deleteAsync instead of removeAsync.

**Comment by vladykin** [comment]
Added uni tests revealed that the CompletionStage returned by deleteAsync actually contains a Boolean (indicating success of deletion), instead of Void (null). This seems to be caused by `com.hazelcast.map.impl.operation.DeleteOperation.getResponse()` returning its success status, which is sent to the client and automatically propagated as result of the CompletionStage.

Should I change deleteAsync return type to `CompletionStage<Boolean>` or find a place to override that Boolean result with `null`?

File: hazelcast/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTest.java
Comment (mdumandag) [reply hazelcast/src/test/java/com/hazelcast/client/map/impl/nearcache/ClientMapNearCacheTest.java:311]: Thanks a lot, I also agree that we don't have to duplicate the tests for the two you wrote

**Comment by mdumandag** [comment]
@vladykin I think it would be good to stick to the CompletionStage<Void> for the return value, to make sure that we are consistent with the blocking version of the method return-value-wise.

**Comment by vladykin** [comment]
Made CompletableFuture returned by `deleteAsync()` always have `null` as result.

Achieved that by modifying `com.hazelcast.map.impl.operation.DeleteOperation.getResponse()` to return `null` instead of `status` (boolean). All tests passed. Are there any consumers of this response which could break after such a change?

BTW, I'm surprised that `new ClientDelegatingFuture<>(future, ss, clientMessage -> null)` didn't do the trick. The 3rd arg `clientMessage -> null`, looked like a good place to transform response to `null`. But that lambda wasn't even invoked when I debugged the failing test.

**Comment by vladykin** [comment]
I addressed all comments from the previous review iteration. Could please someone take a look? Anything else I should fix?

**Comment by mdumandag** [comment]
run-lab-run

**Comment by mdumandag** [comment]
@vladykin can you rebase your branch? It somehow failed to compile but nothing seems off to me.

Regarding the response of the delete operation, I also think that it is not used (as a response), so it seems OK to remove it. 

But I was wondering about your comment about the `ClientDelegatingFuture`. Are you sure that it is not working when the delete operation returns a boolean? I mean, the lambda `clientMessage -> null` is used a lot in the code base, and from the tests I did, it seems like it is working. Can it be the case that the failing test was related to the server-side proxy?

**Comment by mdumandag** [comment]
run-lab-run

**Review by mdumandag — APPROVED** [review]

**Comment by mdumandag** [comment]
@vladykin thanks for the changes, they look good to me. I asked for the required second approval from the core team, and we can merge this PR when they complete their review

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:61]: is this intentional change?

Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:61]: Yes, it was intentional as written in the comments above. It seems we don't need to return a response here, we were not using the response

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:61]: ok makes sense to remove then.

Having a short inline comment for the reasoning of null usage here for future reference will be good.

File: hazelcast/src/main/java/com/hazelcast/client/map/impl/nearcache/NearCachedClientMapProxy.java
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/client/map/impl/nearcache/NearCachedClientMapProxy.java:267]: I'm not sure if this works as expected. The problem is that the call is made async, so the near cache is invalidated before the call has been completed.

Comment (srknzl) [reply hazelcast/src/main/java/com/hazelcast/client/map/impl/nearcache/NearCachedClientMapProxy.java:267]: Note: it is the same for other methods, say putAsyncInternal

Comment (srknzl) [reply hazelcast/src/main/java/com/hazelcast/client/map/impl/nearcache/NearCachedClientMapProxy.java:267]: I don't know why we don't call invalidateNearCache in a whenComplete() call or something like that.

Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/client/map/impl/nearcache/NearCachedClientMapProxy.java:267]: I see that the same pattern is used both in the client and server-side proxies for all invalidating async methods. Since we are invalidating, but not adding anything to the local cache, I don't think it is a problem. At worst, we will remove the key from the local cache, but it won't be removed in the remote if the async call fails. I believe this pattern is fine, as long as we don't add anything to the cache. @ahmetmircik WDYT?

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/client/map/impl/nearcache/NearCachedClientMapProxy.java:267]: @mdumandag Agree that invalidation should be done inside callback after we know that operation is completed; this is the ideal case and can prevent unneeded lookups to remote nodes. An improvement issue can be opened for this. 
BTW we do invalidations in async callbacks for ICache near-cache in similar situations. Aligning  both behaviors will be good also.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java
Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:63]: @ahmetmircik I have committed https://github.com/hazelcast/hazelcast/pull/23509/commits/8b6a1222170860e2de673a61cf204bf05a0bd12f, is this enough?

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:63]: Sure returning null is an option but just want to note my expectation here when i think myself as api user:

Returning a true/false seems more expected api behavior from user perspective. Yes, so far we return nothing to user for sync delete but for async variant, expectations changed and response is passed via callback, i prefer to see the result of api call instead of null.

But it can be only me expecting this, so both case can be ok.

Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:63]: For the member proxy, that is easy to do. It will be a bit problematic for the client-side API. We would need to add the response parameter to the delete codec (that is fine and doable). However, the problem is that when the new client is connected to the old server (pre-5.3), we won't get the response, so we would not have a way of knowing whether it is true or false. If you think it is OK to always return false here (with a bit of documentation), we can do it if everyone agrees on it

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:63]: i'm ok with the approach you proposed. Maybe pre-5.3 can pass null as a response instead of false.

Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:63]: Hmm, that might be a better approach. @vladykin do you want to work on this yourself? if not, I can work on the required protocol/API changes and push them on top of your commits. If you want to do it, basically, you will need to add a new response parameter to this https://github.com/hazelcast/hazelcast-client-protocol/blob/master/protocol-definitions/Map.yaml#L355 (with the since value of 2.6), and update the respective message task and APIs accordingly

Comment (vladykin) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:63]: I'm going to take a look

Comment (vladykin) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:63]: Took a look and now I have several questions to clarify.

1. If I understood the discussion above correctly, the idea is to make the `Delete` operation return true or false, depending on the success of the deletion. Basically, revert my change in the `DeleteOperation.getResponse()`, and instead make the boolean return official by describing it in the protocol spec, something like this. Is that correct?
```yaml
    response:
      params:
        - name: response
          type: boolean
          nullable: true
          since: 2.6
          doc: |
            Returns true if the key exists and removed, otherwise returns false.
```
2. Should the return value of `IMap.delete()` and `IMap.deleteAsync()` in the Java API be changed from void to boolean? 
3. Looks like Delete operation doesn't always know if the value was actually deleted. In the current implementation, if the value is not cached in memory, but exists in the MapStore, then the return value is `false`. Is that ok?
4. That's a separate git repository. What's the proper way to make related changes to two repositories?

Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:63]: 1. Yes, correct. Just make the `nullable: false`please, others look good.
2. Yes.
3. @ahmetmircik what do you think?
4. You do your changes in the hazelcast-client-protocol locally, generate codecs and add them to this PR. Then, you create a pull request on the hazelcast-client-protocol repo. After merging this PR, we will merge the protocol PR as well.

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:63]: Regarding 3, we can keep what we have currently, if key is in memory and deletion is successful, return true otherwise return false.

Comment (vladykin) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:63]: AFAIK, changing the return type of an existing API method [is not a binary-compatible change](https://www.morling.dev/blog/refining-return-type-java-methods-without-breaking-backwards-compatibility/). That means, every dependent piece of code, that uses `IMap.delete()`, will break, when switched to the new version of `hazelcast.jar`. To avoid breaking, the code will have to be recompiled from source.

Is that acceptable? What are the backward compatibility requirements to the Hazelcast library?

Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java:63]: No, we cannot change the return type of the `void delete()` method as you said. We were talking about making the new API return a `CompletionStage<Boolean>`, instead of `CompletionStage<Void>`

Comment (vladykin) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/DeleteOperation.java]: Done. Regenerated Java code with the updated protocol (1st commit) and added `CompletionStage<Boolean> deleteAsync(K key)` (2nd commit).

The build was failing for me locally. Enforcer plugin complained on the bytecode version in netty 4.1.90. I had to revert the recent bump of netty version to make the build pass. Hope someone will soon land a proper fix for this to the master branch, so I can remove this hack from the PR

**Comment by mdumandag** [comment]
run-lab-run

**Comment by vladykin** [comment]
Rebased on the current master. Removed the revert for netty version bump.

**Comment by mdumandag** [comment]
run-lab-run

**Comment by vladykin** [comment]
Rebased on the current master. Created a separate PR for the protocol change: https://github.com/hazelcast/hazelcast-client-protocol/pull/466

**Comment by vladykin** [comment]
Manually rebased an resolved a conflict in binary files generated from client protocol. There was another protocol change landed recently, and git can't merge changes in binary files, even though the protocol changes are not conflicting

**Comment by vladykin** [comment]
@mdumandag, @ahmetmircik, could you please see if I need to do anything else before you can accept this pull request?

**Comment by vladykin** [comment]
@mdumandag, @ahmetmircik, looks like this PR has been waiting for Hazelcast-pr-builder for more than a month. Could you please trigger it?

**Comment by vladykin** [comment]
@ihsandemir, thanks for triggering the build!

I see it failed. Unfortunately, the failure message above doesn't include any useful information. Clicking on "Click to expand the log file" expands an empty box.

The links to [build log](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/19073/Hazelcast-pr-builder-19073.zip) and [artifacts](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/19073/Hazelcast-pr-builder-19073.zip) are not accessible to me.

Could you please let me know, what the failures are?

I rebased my patch and ran `mvn clean package` locally. A couple of tests failed: a few Mongo-related tests in `hazelcast-sql` returned wrong results and two in `hazelcast-jet-python` timed out. Errors look completely unrelated to my change. Are all tests passing in the master branch now?

**Comment by ihsandemir** [comment]
@vladykin Unrelated test `com.hazelcast.internal.tpcengine.nio.NioAsyncServerSocketTest.test_createCloseLoop_withSameReactor` seems to be failed. I will re-run the tests.

```
java.io.UncheckedIOException: Failed to bind to /127.0.0.1:5001
	at com.hazelcast.internal.tpcengine.nio.NioAsyncServerSocket.bind(NioAsyncServerSocket.java:106)
	at com.hazelcast.internal.tpcengine.net.AsyncServerSocket.bind(AsyncServerSocket.java:102)
	at com.hazelcast.internal.tpcengine.net.AsyncServerSocketTest.test_createCloseLoop_withSameReactor(AsyncServerSocketTest.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:49)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:120)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:95)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:69)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
Caused by: java.net.BindException: Address already in use
	at java.base/sun.nio.ch.Net.bind0(Native Method)
	at java.base/sun.nio.ch.Net.bind(Net.java:459)
	at java.base/sun.nio.ch.Net.bind(Net.java:448)
	at java.base/sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:227)
	at com.hazelcast.internal.tpcengine.nio.NioAsyncServerSocket.bind(NioAsyncServerSocket.java:104)
	... 43 more
```

**Comment by ihsandemir** [comment]
run-lab-run

**Comment by ihsandemir** [comment]
See existing [issue](https://github.com/hazelcast/hazelcast/issues/24002) for the unrelated failure.

**Comment by vladykin** [comment]
@ihsandemir, thank you! The new build passed, all checks in this PR are green. Can it be merged now?

**Comment by vladykin** [comment]
@ihsandemir, @mdumandag, @ahmetmircik, I noticed that the status of this PR changed from "Ready to merge" to "Code owner review required". I haven't seen any notification about that. Has someone withdrawn their prior approval? Are there any questions/concerns I should address?

**Review by ahmetmircik — APPROVED** [review]

**Pull request merged**
